Framework: Add Component Storybook.

epiqueras opened the pull request:
## Description

This PR adds [Storybook](https://storybook.js.org/) support to the repo and uses it to create a storybook for `@wordpress/components` using the `Button` component as an example.

It also adds plugins for automatically running and displaying accessibility tests right besides stories, simulating color blindness, and changing the viewport size/device.

The PR also modifies the Travis config so that the storybook is deployed to GitHub Pages using the `/design-system/components` path.

This gave me the idea that maybe the GitHub Pages page should use a custom domain in `wordpress.org` and present some sort of landing page for the JS focused WordPress efforts. That can then link to the editor playground and the storybook.

## How has this been tested?

The new storybook was verified to work in development (`npm run design-system:dev`) and production (`npm run design-system:build`) modes.

## Screenshots

#### Passing a11y Tests

<img width="1669" alt="Screen Shot 2019-09-18 at 1 49 58 PM" src="https://user-images.githubusercontent.com/19157096/65188598-a2b51e00-da23-11e9-9d78-57822108ca89.png">

#### Failing a11y Tests

<img width="1673" alt="Screen Shot 2019-09-18 at 1 51 04 PM" src="https://user-images.githubusercontent.com/19157096/65188608-a9439580-da23-11e9-8dbf-86f029800932.png">

#### Color Blindness Simulator

<img width="636" alt="Screen Shot 2019-09-18 at 2 51 57 PM" src="https://user-images.githubusercontent.com/19157096/65188714-e60f8c80-da23-11e9-9c72-32963b5b9ee8.png">

#### Changing Viewports

<img width="1148" alt="Screen Shot 2019-09-18 at 2 15 13 PM" src="https://user-images.githubusercontent.com/19157096/65188630-b496c100-da23-11e9-82c0-6c9ccaf15a52.png">

## Checklist:

- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

epiqueras committed:
Framework: Add Component Storybook.

gziolo commented:
How would it play with #17173 from @itsjonq where he proposes a site with live examples of components?

There is also the [playground](https://github.com/WordPress/gutenberg/tree/master/playground).

In addition, there is also a proposal to create some specialized testing app for RichText proposed by @ellatrix in https://github.com/WordPress/gutenberg/pull/17303.

Moving forward, we will also need playgrounds for iOS and Android example apps working with React Native. See https://github.com/WordPress/gutenberg/pull/17456 from @Tug.

I guess, all individual solutions makes sense but we should coordinate everything and bring them as close as possible so we don't have to maintain 6 completely independent apps with their own unique tooling and build steps. Do you have any thoughts on how this could be consolidated as much as possible?

Should we start with some sort of proposal which describes the role of every tool and how they can benefit workflows of different types of contributors?

/cc @jameslnewell @griffbrad @youknowriad

nb commented:
Thanks, @epiqueras – storybook has improved since the last time I looked at it. The keyboard shortcuts are handy.

Is there a chance we can get as part of the netlify deploy? Not having to run it locally will be super useful. I thought twice before running it locally :)

The in-browser test also look like a useful tool.

On styles – the link uses the default browser styles, do you think it would make sense to put it in some WordPress style context?

How was the story format chosen? Both mdx and `storiesOf` look interesting.

And few additions that would be helpful, even if not in this PR:
- A way to see the source code of the story.
- Be able to change props on the fly. I liked the [“knobs” tab in this example](https://uber.github.io/react-vis/website/dist/storybook/).
- Have a link to the documentation for each component.

epiqueras committed:
Framework: Add storybook to Netlify deploy.

epiqueras commented:
@gziolo 

>How would it play with #17173 from @ItsJonQ where he proposes a site with live examples of components?

This is a much simpler alternative approach. We should however, theme it so that it looks more like WordPress docs, but that can be a different PR.

>There is also the playground.
>
>In addition, there is also a proposal to create some specialized testing app for RichText proposed by @ellatrix in #17303.

Yeah, those are parallel things.

>Moving forward, we will also need playgrounds for iOS and Android example apps working with React Native. See #17456 from @Tug.

Yeah, Storybook can support React Native.

>I guess, all individual solutions makes sense but we should coordinate everything and bring them as close as possible so we don't have to maintain 6 completely independent apps with their own unique tooling and build steps. Do you have any thoughts on how this could be consolidated as much as possible?
>
>Should we start with some sort of proposal which describes the role of every tool and how they can benefit workflows of different types of contributors?

This is a very conservative start. I'm making no assumptions about any sort of decision beyond using Storybook for the component playground.


@nb 

>Thanks, @epiqueras – storybook has improved since the last time I looked at it. The keyboard shortcuts are handy.

Yeah it looks like it reclaimed its throne in the UI playground space. They even support MDX docs now.

>Is there a chance we can get as part of the netlify deploy? Not having to run it locally will be super useful. I thought twice before running it locally :)

Yeah me too. I just added it.

>On styles – the link uses the default browser styles, do you think it would make sense to put it in some WordPress style context?

I'm not sure. I added the Gutenberg styles, but if this will be potentially used outside of a WordPress context, maybe it's a sign we need to add more default styles?

>How was the story format chosen? Both mdx and storiesOf look interesting.

Good question.

The `storiesOf` is the old, very clunky API, which is no longer recommended.

MDX is for authoring freeform Docs with embedded stories like React Styleguidist. It can and should complement the playground.

>A way to see the source code of the story.

The addon that does that is out of date, so it might be a bit involved, but another PR can take care of it.

>Be able to change props on the fly. I liked the “knobs” tab in this example.

I almost added the Knobs addon, but linking props to it is a very manual process and I want to be conservative about enforcing things on contributors. If we think it's worth it, we can add it.

>Have a link to the documentation for each component.

Yeah, that's what MDX will be for.

epiqueras commented:
https://5d83938b1a0a5b000a81d1f1--gutenberg-playground.netlify.com/design-system/components

ItsJonQ commented:
Just took a look! @epiqueras thank you for putting this together! I think Storybook is a fantastic addition to the UI dev/design work flow (essential IMO).

@gziolo Re: documentation

@epiqueras shared this link with me:
https://medium.com/storybookjs/storybook-docs-sneak-peak-5be78445094a

If we're able to get this into Storybook, then it is able to accommodate the needs that the doc site (that I had proposed) attempted to address.

cc'ing @jameslnewell 😊

epiqueras committed:
Framework: Add Storybook Docs.

epiqueras committed:
Framework: Add a stand-alone Storybook documentation page example.

epiqueras commented:
I've gone ahead and added the new MDX doc features.

Every story file now gets a docs page automatically generated, but you can overwrite them with custom MDX files.

I've also added a stand-alone documentation page as an example.

I think this new "Docs" feature suite removes the need for most of the plugins talked about above. We can always add more later on PRs that actually make real use of them.

Here is more information on the new features:

https://github.com/storybookjs/storybook/blob/next/addons/docs/README.md#more-resources

Take a look at how pages are automatically generated, how MDX can be used to overwrite them, and how you can theme the docs and export them on their own.

epiqueras commented:
https://deploy-preview-17475--gutenberg-playground.netlify.com/design-system/components/?path=/docs/introduction--page

diegohaz commented:
@epiqueras Could you try to create stories for a component that has JSX?

I'm using your PR to create stories for `Toolbar`, but seeing this error on the Storybook page:
```
ReferenceError: React is not defined
    at Module../packages/components/src/popover/index.js (http://localhost:50417/main.b1ea0813c451b7179bb3.bundle.js:3941:1)
    at __webpack_require__ (http://localhost:50417/runtime~main.b1ea0813c451b7179bb3.bundle.js:786:30)
    at fn (http://localhost:50417/runtime~main.b1ea0813c451b7179bb3.bundle.js:151:20)
    at Module../packages/components/src/tooltip/index.js (http://localhost:50417/main.b1ea0813c451b7179bb3.bundle.js:5884:66)
    at __webpack_require__ (http://localhost:50417/runtime~main.b1ea0813c451b7179bb3.bundle.js:786:30)
    at fn (http://localhost:50417/runtime~main.b1ea0813c451b7179bb3.bundle.js:151:20)
    at Module../packages/components/src/icon-button/index.js (http://localhost:50417/main.b1ea0813c451b7179bb3.bundle.js:2903:66)
    at __webpack_require__ (http://localhost:50417/runtime~main.b1ea0813c451b7179bb3.bundle.js:786:30)
    at fn (http://localhost:50417/runtime~main.b1ea0813c451b7179bb3.bundle.js:151:20)
    at Module../packages/components/src/toolbar-button/index.js (http://localhost:50417/main.b1ea0813c451b7179bb3.bundle.js:5463:70)
```

It's fixed if I add an empty `.babelrc` file to `packages/components/storybook`, but it's weird anyway! 🙃

epiqueras committed:
Components: Add custom Babel config for storybook.

epiqueras commented:
>It's fixed if I add an empty .babelrc file to packages/components/storybook, but it's weird anyway!

@diegohaz It was picking up the root babel config and that transforms JSX into calls to `wp.element.createElement`. I've added the specific `.babelrc` file and it works fine now.

Thanks!

gziolo commented:
> It's fixed if I add an empty `.babelrc` file to `packages/components/storybook`, but it's weird anyway! 🙃

It seems like Storybook performs some magic operations on Babel config, similar to what Parcel does. This is what we had to add for Parcel:
https://github.com/WordPress/gutenberg/blob/bf1ca0771f655121a569a8b2776768b5d0e0fd17/playground/.babelrc#L1-L9

youknowriad commented on package.json:
```diff
@@ -218,10 +223,12 @@
 		"preenv": "npm run check-engines",
-		"env": "wp-scripts env"
+		"env": "wp-scripts env",
+		"design-system:dev": "concurrently \"npm run dev:packages\" \"start-storybook -c ./packages/components/storybook\"",
```
for playground we use `:start` and for gutenberg we use `:dev` I personally prefer start but we should try to consolidate all these similar commands on `:dev` maybe to avoid impacting the plugin devs.

epiqueras replied:
I've seen `start` used more to start production apps like web servers, so I kind of liked we opted for `dev`.

youknowriad replied:
Can we update the playground one?

epiqueras replied:
Done

youknowriad commented on packages/components/docs/introduction.story.mdx:
```diff
@@ -0,0 +1,7 @@
+
+# Introduction
+
+Hello World
```
Are we concerned about the duplication in terms of docs, this will introduce. Components have READMEs that are currectly synced into the WordPress handbook and they'll have special docs that will live in the storybook?

epiqueras replied:
These are the design system docs.

A lot of that content will probably be moved here though.

youknowriad replied:
> These are the design system docs.

This is what the README are for. so the components part for the design system will get duplicated?

epiqueras replied:
Yes, but MDX is much better than a simple README.md and will be the new framework for them.

I guess we can delete the READMEs once the new docs are fleshed out.

youknowriad replied:
We can't delete the READMEs until we figure out how to show Storybook in the wp.org devdocs. I'm concerned that it will take a lot of time and that we'll have to maintain duplicated content for a long time.

epiqueras replied:
Storybook can export static files, that won't be hard to do. And it won't be duplicated, the new docs will be better and there won't necessarily be a 1 to 1 mapping, because the goal is to overhaul this package into a proper design system component kit.

youknowriad commented on packages/components/src/button/stories/index.js:
```diff
@@ -0,0 +1,41 @@
+/**
```
Is it possible to move it out of the `components` package?

epiqueras replied:
It can be wherever, but I don't think that's a good idea. These are like tests and should live next to components.

Other storybooks, like the block editor's, might warrant a different file structure though.

youknowriad replied:
I can see that but if we're using Storybook as a doc website, I think it defeats that argument. We should first clarify what we want Storybook to be used for. 

If it's for a docs website, I believe a separate location is better, if it's just for having stories to develop the component, I think having them in subfolders is fine.

I can also see a folder structure like that:

```
playground
   - design-system
   - block-editor (current playground) 
    - ios (these exist today but in the mobile repo)
    - android
```

basically a folder for all playgrounds.

epiqueras replied:
We are using it for both. Storybook has a component playground mode and a docs website mode.

The playground mode files should live next to their components, because they are like tests and will need to reference them heavily. It also lets you easily see which components are missing stories.

The docs mode files should live in a docs folder in the root of their package, like `docs/introduction.story.mdx` here, because they are just freeform documentation and are not tied to a specific component.

youknowriad replied:
I feel we're trying to solve a lot at the same time. While I agree that it would be great to use storybook for both, I think we should have a clear plan about how to update the docs, how to sync them to wp.org docs. Right now when I look at the output, it's a mix of both which I'm not sure is perfect to sync as is into wp.org. Also how to do it.

I'd be ok to land storybook as an experiment for now until we figure out all these things but an experiment shouldn't impact the current codebase a lot that's why I wonder if it's better to start as a separate folder

epiqueras replied:
>I feel we're trying to solve a lot at the same time.

We are just solving "choosing the docs framework for reusable components".

>I think we should have a clear plan about how to update the docs, how to sync them to wp.org docs.

Travis job that deploys static files.

>Right now when I look at the output, it's a mix of both which I'm not sure is perfect to sync as is into wp.org

There's CLI options for separating the 2 or we can theme it so that it fits better into .org's site patterns.

>I'd be ok to land storybook as an experiment for now until we figure out all these things but an experiment shouldn't impact the current codebase a lot that's why I wonder if it's better to start as a separate folder

What could change is how we theme and layout Storybook's output, but not the use of Storybook, it's by far the best option out there for this.

Right now this just uses defaults for everything, but with this as a base, designers can lay things out with custom navigation, layouts, colors, and whatever else the .org site needs.

epiqueras commented:
>It seems like Storybook performs some magic operations on Babel config, similar to what Parcel does. This is what we had to add for Parcel:

Yeah, we don't need that though, because Storybook uses its own version of React.

youknowriad reviewed (approved):
I'm willing to merge this but we need a plan to move this forward, should we create a follow-up issues about adding stories for all components, how about docs? who own this?

epiqueras committed:
Framework: Update lockfile and playground script name.

epiqueras commented:
>I'm willing to merge this but we need a plan to move this forward, should we create a follow-up issues about adding stories for all components, how about docs? who own this?

cc @ItsJonQ 

Merging as soon as Travis is finished here :tada:

gziolo reviewed (changes requested):
The lock file should be updated. I did some cleanup yesterday and it looks like this PR would revert it. See https://github.com/WordPress/gutenberg/pull/17705 and the related part:


> In addition, it follows-up #17668 and does some cleanup with dependencies of the WP packages to ensure that all of them are properly hoisted to the root of `node_modules`. We should consider adding some check to Travis which is able to detect whether `node_modules` folder was created inside andy of the `packages` subfolders.

gziolo commented:
With the current location of Storybook, the list of files ignored by Lerna should be updated:
https://github.com/WordPress/gutenberg/blob/a706b426ece5cfa5b5e2ed4c84d75a4d95780e5a/lerna.json#L7-L11

There is also this question of whether the whole folder of Storybook should be published to npm. I guess it should be filtered out. In some packages, we whitelist what's sent to npm.

epiqueras committed:
Framework: Update lockfile and ignore Storybook files for publishing.

epiqueras commented:
@gziolo https://github.com/WordPress/gutenberg/pull/17475/commits/33e436961db8e91e8683fb3b62235fe27ad77d04

Does that look better?

>There is also this question of whether the whole folder of Storybook should be published to npm. I guess it should be filtered out. In some packages, we whitelist what's sent to npm.

It's ignored by lerna and not included in the build. Are we somehow including all files by default?

gziolo commented on .travis.yml:
```diff
@@ -201,6 +201,7 @@ jobs:
 before_deploy:
   - npm install
   - npm run playground:build -- --public-url '/gutenberg'
+  - npm run design-system:build
```
This will run on for all jobs when merging into `master`. We already have this issue for the playground. It will get only worse:

![Screen Shot 2019-10-04 at 07 38 25](https://user-images.githubusercontent.com/699132/66183707-0e30f980-e67a-11e9-9b44-c6d3466e46af.png)

We will need to improve the setup of Travis to run deploys only once per PR merged.

epiqueras replied:
We need two stages. I'll fix it.

gziolo commented:
> It's ignored by lerna and not included in the build. Are we somehow including all files by default?

Lerna uses those globs to ignore matching files from version update. It doesn't interact with how files are published to npm. By default, npm publishes all files in the package.

gziolo reviewed (approved):
I opened #17762 which contains several tweaks to this setup. It'd be great to land them together with this PR. Awesome work integrating Storybook.

I love that it can be run in the local network with easy making testing experience so great ❤️

gziolo commented:
Huh, a conflict in the lock file, probably after I landed webpack upgrade. I will fix it.

gziolo commented:
I rebased changes with `master` and regenerated the lock file again in d491df8. It looks like the lock file changes are smaller now. I suspect it's because we are using the latest version of `webpack` now.

gziolo commented:
Unit tests might fail from time to time so it would be nice to address this comment from my accompanied PR:

https://github.com/WordPress/gutenberg/pull/17762#issuecomment-538337715

epiqueras committed:
Travis: Only deploy once per build, not per job.

epiqueras committed:
Framework: Update lockfile.

gziolo commented:
https://wordpress.github.io/gutenberg/ is broken, not sure if this is related to this PR but we need ot investigate it.

gziolo commented on .travis.yml:
```diff
@@ -195,18 +195,21 @@ jobs:
+      before_deploy:
+        - npm install
+        - npm run playground:build -- --public-url '/gutenberg'
+        - npm run design-system:build
```
It looks like with the changes applied in this PR, `design-system:build` will run twice on Travis:
```json
"playground:build": "npm run build:packages && parcel build playground/src/index.html -d playground/dist && npm run design-system:build",
```

epiqueras replied:
#17818

epiqueras commented:
@gziolo 
>https://wordpress.github.io/gutenberg/ is broken, not sure if this is related to this PR but we need ot investigate it.

#17818 also fixes this.
